mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ExecutorStateManagerImpl.java

```diff
@@ -227,10 +251,56 @@ public Optional<Entry<TaskExecutorID, TaskExecutorState>> findFirst(
     }
 
     @Override
-    public Optional<Pair<TaskExecutorID, TaskExecutorState>> findBestFit(TaskExecutorAssignmentRequest request) {
+    public void unscheduleJob(String jobId) {
+        pendingJobRequests.invalidate(jobId);
+    }
+
+    @Override
+    public Optional<BestFit> findBestFit(TaskExecutorBatchAssignmentRequest request) {
+
+        if (request.getAllocationRequests().isEmpty()) {
+            log.warn("TaskExecutorBatchAssignmentRequest {} with empty allocation requests.", request);
+            return Optional.empty();
+        }
+
+        boolean noResourcesAvailable = false;
+        final Map<TaskExecutorAllocationRequest, Pair<TaskExecutorID, TaskExecutorState>> bestFit = new HashMap<>();
+        final boolean isJobIdAlreadyPending = pendingJobRequests.getIfPresent(request.getJobId()) != null;
+
+        for (Entry<MachineDefinition, List<TaskExecutorAllocationRequest>> entry : request.getGroupedByMachineDef().entrySet()) {
+            final MachineDefinition machineDefinition = entry.getKey();
+            final List<TaskExecutorAllocationRequest> allocationRequests = entry.getValue();
+
+            Optional<Map<TaskExecutorID, TaskExecutorState>> taskExecutors = findTaskExecutorsFor(request, machineDefinition, allocationRequests, isJobIdAlreadyPending);
+
+            // Mark noResourcesAvailable if we can't find enough TEs for a given machine def
+            if (!taskExecutors.isPresent() || noResourcesAvailable) {
+                noResourcesAvailable = true;
+                continue;
```

Comment:
nit: u can just break here

Reply:
I don't think breaking is the right thing to do in here because we can potentially have more than 1 sku to add to the pending list so we want to loop through all the entries. Let me give you an example: job A has stage 1 with sku small and stage 2 with small medium. If both don't have enough resources available we want to insert both into the pending set which happens on `findTaskExecutorsFor`; however if we break after finding out that there are not enough resources available on sku small we would miss out on adding the medium sku into the pending set.

Reply:
you're right. this was there because of an older implementation of the cache.